Add l7fw app and `snabb wall filter`

takikawa opened the pull request:
This PR adds a `l7fw` app and a `snabb wall filter` command.

@aperezdc, @dpino : would be interested in any feedback you have, particularly at a high-level to see if the implemented features look sufficient. Here are some design points that maybe could use feedback:

  * Supporting pfmatch actions that can reference the number of packets in a given protocol flow required changing the pflua library in a somewhat invasive way. Does changing pflua seem okay, or should I just remove this feature?
  * So far `snabb wall filter` is mostly for testing and I was assuming someone who wanted a firewall would arrange the `l7spy` and `l7fw` apps themselves. Should `snabb wall filter` also have an option to output directly to a NIC instead?
  * Are there any other features needed? (one that I plan to add before merging is logging actions to a plaintext file, potentially with different levels of verbosity)

The firewall currently accepts as its rules a table like the following:

```
{ DNS = "accept",
  SSL = "drop",
  HTTP = [[match { flow_count > 2 => reject; otherwise => accept }]] }
```

where the keys are protocol names and the values are firewall actions (the third one is a pfmatch expression). I plan to document this in high-level user docs. This PR includes API docs for the included apps, but no high-level docs yet.

The `snabb wall filter` command that's implemented lets you run an `l7fw` and `l7spy` app together to filter packets from any source that `snabb wall spy` accepts. Right now it just supports sending accepted packets and reject responses to pcap files.

dpino commented on lib/pflua/src/pf/utils.lua:
```diff
@@ -11,6 +11,11 @@ struct pflua_timeval {
+-- Additional function arguments other than 'len' and 'P' that the
+-- compiled function may accept (this needs to be tracked in several
+-- pflua passes, which is why the data is kept here)
+extra_args = {}
```
Some reflections about the `extra_args` array.

I like it but it's true that is a bit invasive. To minimize this effect, would it make sense to treat "len" as a default extra_arg? What I see is that now whenever there is a check about whethen a token is "len", there's an extra check to see if the token is in `extra_args` too. If "len" is a default extra_arg, we could remove one check.

Another thing I'm not sure about. Would it make sense to store `extra_args` as a global so it's not necessary to reference utils? (could be localized as well per module). For instance:

```lua
_G["extra_args"] = {}
```

The inconvenient is that if pflua is used in combination with another library that has this global defined there might be overrides.

aperezdc replied:
I certainly would rather not pollute the global namespace. Given that the value is a table, the variable can be localized *as long as it is never reasiigned* (the table items can be cleared, though). That would be okay to avoid always referencing it as `utils.extra_args`, adding a comment next to the definition explaining that it shouldn't be reassigned.

takikawa replied:
Generalizing this to include `len` is a good idea, thanks. I implemented the change and then localized the variable in each module.

dpino commented on src/apps/wall/l7fw.lua:
```diff
@@ -0,0 +1,275 @@
+-- called by pfmatch handlers, just drop the packet on the floor
+function L7Fw:drop(pkt, len)
+   packet.free(self.current_packet)
+   return
```
Why the extra return? AFAIK, `return` is the same as `return nil`. If a Lua function has no return statement, returns nil. So, it seems to me this `return` could be omitted.

takikawa replied:
I think the `return` may be a leftover from an earlier draft. Either way, I removed it.

dpino commented on src/apps/wall/l7fw.lua:
```diff
@@ -0,0 +1,275 @@
+
+-- called by pfmatch handler, handle rejection response
+function L7Fw:reject(pkt, len)
+   link.transmit(assert(self.output.reject,
```
I think it's better to move the assert to `L7Fw:push()`. I explain why (let me know if it makes sense to you).

`L7Fw:reject()` is only called from `L7Fw:push()`. Imagine you setup an app graph without a reject link and then because of the packet flow you got a reject. The execution of the program will break while running. As for me, I would prefer the program to break as soon as possible. That's why I'd move the assert to `L7Fw:push()`.

takikawa replied:
Yes, I think that makes sense. I put the `assert` there so that having a `reject` link could be optional if that kind of rule isn't used. But it probably makes more sense to make it mandatory in all cases.

dpino commented on src/apps/wall/l7fw.lua:
```diff
@@ -0,0 +1,275 @@
+   local is_tcp  = false
+   local ip_protocol
+
+   if ip_orig:version() == 4 then
```
Maybe is worth to write these values as named constants such as PROTO_IPV4, PROTO_IPV6, etc, like in pflua.

aperezdc replied:
Side note: Many of the constants from `scanner.lua` are likely to be similar or the same to the ones in `src/apps/lwaftr/constants.lua` — later on, when working on upstreaming SnabbWall, IMHO it would make sense to try and merge both.  There are constants readily available at `src/apps/wall/scanner.lua`that maybe it's reasonable to reuse. Best thing is probably to move them into `src/apps/wall/constants.lua` and `require()` that from both the `scanner.lua` and `l7fw.lua`.

---

Side note: Many of the constants from `scanner.lua` are likely to be similar or the same to the ones in `src/apps/lwaftr/constants.lua` — later on, when working on upstreaming SnabbWall, IMHO it would make sense to try and merge both.

takikawa replied:
In my updated commits I made some of the constants (the ones used several times) into named definitions. I thought about making a new constants module, but it didn't seem like it would be worth it. (especially if these constants are shared with `lwaftr` later as suggested)

dpino commented:
> Supporting pfmatch actions that can reference the number of packets in a given protocol flow required
> changing the pflua library in a somewhat invasive way. Does changing pflua seem okay, or should I just 
> remove this feature?

The feature is worth to me. It's true that it's a bit invasive. If we can minimize this effect, it would be ideal. It gives pflua more possibilities to be extended in the future.

> So far snabb wall filter is mostly for testing and I was assuming someone who wanted a firewall would 
> arrange the l7spy and l7fw apps themselves. Should snabb wall filter also have an option to output directly 
> to a NIC instead?

I think not. To me is fine as it's now. Someone interesting in a firewall app can still use L7Fw, which contains the interesting logic. `snabb wall filter` is mostly UI and it works as fine an example of how to use L7Fw. There are other programs in snabb codebase that are simply there as examples, for instance the `example_spray` or `example_replay` programs. I see "snabb wall filter" as a similar case.

> Are there any other features needed? (one that I plan to add before merging is logging actions to a plaintext 
> file, potentially with different levels of verbosity)

Sounds interesting. If you feel like implementing this feature, I'd put it under a debug or verbosity flag (printing to stdout or a file may decrease speed significantly). At this moment, I cannot see more features.

takikawa commented:
Thanks for the comments. I think I've addressed them in the new commits I pushed.

I also implemented a simple `report` method for `l7fw` for testing and added support for (optional, disabled by default) logging to the system log. I made the logging options similar to those for the `ufw` firewall (just `"off"` and `"on"` for now, but it could support levels like `"low"`, `"high"`, etc. like `ufw` later).

One issue with merging this though is that there are some commits that should probably go upstream separately, like 3fd40e8 (merged in upstream pflua but not in Snabb),  7417375 (in [this PR](https://github.com/snabbco/snabb/pull/1095)), and 0e692d5 (included with other pflua changes not yet merged upstream) that are needed to run the new app/program. Any suggestions for what to do there?

(For example, I could just merge them into the main snabbwall repo, and then later just deal with a potential merge conflict when the identical change is merged from the upstream snabb release)

dpino commented:
@takikawa I generally I do what you did. If I found fixes that could go to upstream, I send a PR with the individual commit and I keep the commit in my branch too.

So now this will go to Adrián's snabbwall branch. The conflicts will happen once snabb:master is merged into Adrián's snabbwall. I dunno if latest versions of git would detect the commits are duplicated and discard them, but in case not solving them manually sounds like a good approach to me. I don't know anything better :)

dpino commented on src/apps/wall/l7fw.lua:
```diff
@@ -0,0 +1,353 @@
+   print(("Dropped packets:  %d (%d%%)"):format(dropped, d_pct))
+end
+
+local logging_priority = bit.bor(LOG_USER, LOG_INFO)
```
BitOps module is always accessible in LuaJIT but the docs recommend to explicitly require it:

> The suggested way to use the BitOp module is to add the following to the start of every
> Lua file that needs one of its functions:
> local bit = require("bit")
> This makes the dependency explicit, limits the scope to the current file and provides faster access to the bit.*

Source: http://bitop.luajit.org/api.html

This is more of a nit but I think it's helpful LuaJIT knowledge too.

aperezdc commented on src/apps/wall/README.md:
```diff
@@ -12,7 +12,56 @@ L7Spy (apps.wall.l7spy)
+an instance of the `Scanner` class. The scanner instance may be shared among
+several `L7Spy` instances or with a `L7Fw` app for filtering.
+
+— Method **L7Fw:new** *config*
```
This should be ``L7Spy:new``.
